classifier_and_summarizer/summarizer/budgeted_max_cover_solver.py

```diff
@@ -0,0 +1,256 @@
+          cover_index : index of the cover in the list_of_covers
+
+        Returns : an integer representing the number of elements covered by
+        the cover_picked but not by an previously chosen cover
+        '''
+        weight = 0
+        cover_picked = self.list_of_covers[cover_index]
+        for i in range(self.number_of_elements):
+            if cover_picked.cover[i] == 1 and self.is_element_picked[i] == 0:
+                weight += 1
+        return weight
+
+    def cost_for_picking_element(self, cover_index):
+        ''' return the cost of the cover'''
+        return self.list_of_covers[cover_index].cost
+
+    def comparator_for_sorting(self, cover_index):
+        '''Provides a simple comparator for sorting the unpicked covers'''
+        return max(1, self.weight_for_uncovered_elements(cover_index)) \
+            / self.cost_for_picking_element(cover_index)
+
+    def initialize_auxiliary_variables(self):
+        '''initializes the auxiliary variables'''
+        self.cost_for_covers_picked_so_far = 0
+        self.number_of_elements_picked = 0
+        self.covers_picked_so_far = []
+        self.is_element_picked = [0]*self.number_of_elements
+
+    def sum_of_costs_for_covers(self, covers_picked):
+        ''' returns the total cost of a given set of covers'''
+        return sum(self.list_of_covers[i].cost for i in covers_picked)
+
+    def cover_can_be_picked(self, cover_index):
+        ''' returns True if cover can be picked for given budget'''
+        cost_if_picked \
+            = self.list_of_covers[cover_index].cost \
+            + self.cost_for_covers_picked_so_far
+
+        is_not_picked \
+            = self.list_of_covers[cover_index].id not in \
+            self.covers_picked_so_far
+
+        return cost_if_picked <= self.budget and is_not_picked
+
+    def find_max_cover_given_initial_covers(self, initial_covers_picked):
+        '''
+        Finds the maximum cover given a starting cover.
+
+        Given an initial list of covers to start with,
+        iteratively expands the list of covers to cover
+        more elements within the given budget
+        Params:
+          initial_covers_picked = the initial list of covers we start with
+
+        Returns : attriutes for the best set of covers returned as a dict
+        '''
+        self.initialize_auxiliary_variables()
+        for cover in initial_covers_picked:
+            self.pick_cover(cover)
+
+        some_cover_can_be_picked = True
+
+        while some_cover_can_be_picked:
+
+            candidate_cover_indices = [i for i in range(self.cover_size)
+                                       if
+                                       self.cover_can_be_picked(cover_index=i)]
+
+            candidate_cover_indices.sort(
+                key=self.comparator_for_sorting, reverse=True)
+
+            some_cover_can_be_picked = True
+
+            if candidate_cover_indices:
+                self.pick_cover(candidate_cover_indices[0])
```

Comment:
probably we can use the python min() function, and pass in the key function as the weight/cost ratio. This should shorten this part, something on the lines of:

while True:
   candidate_cover_indices = [...]
   if len(candidate_cover_indices):
     break
   self.pick_cover(min(candidate_cover_indices, key_function))

Reply:
Thanks for the suggestion,
I'd have to use max function since i need the cover with the maximum weight:cost ratio
Will modify accordingly.